byron/ledger/executable-spec/src/Ledger/UTxO.hs

```diff
@@ -110,76 +120,96 @@ instance STS UTXO where
     [ return (UTxO Map.empty) ]
   transitionRules = [ utxoInductive ]
 
-utxoInductive :: TransitionRule UTXO
+utxoInductive :: Ord id => TransitionRule (UTXO id)
 utxoInductive = do
-  TRC (pc, utxo, tx) <- judgmentContext
+  TRC (UTxOEnv _ pc, utxo, tx) <- judgmentContext
   balance (txouts tx)
     <> txfee tx
     == balance (txins tx ◁ utxo)
     ?! IncreasedTotalBalance
   pcMinFee pc tx <= txfee tx ?! FeeTooLow
-  let unspentInputs (UTxO aUtxo) = Map.keysSet aUtxo
-  txins tx `Set.isSubsetOf` unspentInputs utxo ?! BadInputs
+  let
+    unspentInputs (UTxO aUtxo) = Map.keysSet aUtxo
+    txinsSet = Set.fromList $ txins tx
+  Set.size txinsSet == length (txins tx)
+    && txinsSet `Set.isSubsetOf` unspentInputs utxo
+    ?! BadInputs
```

Comment:
I guess when running the tests we would like to show the reason why inputs are considered "bad", but that would make it impossible for `applySTS` to be used inside the `hedgehog` `Update` constructor. Does that mean that we have to remove all the information from the other predicate failures if we want to rely on references for shrinking?

Reply:
We can show it as long as we don't have to show the `id`, so we just have to write the `Show` instances manually

Reply:
Yeah, so we won't be able to show much if we can't show the `id`'s, right?

Reply:
Oh, good point, so I guess we can wait until some form of `GADT`s get added to `hedgehog`